src/gui/application.cpp

```diff
@@ -674,4 +675,20 @@ void Application::slotGuiIsShowingSettings()
     emit isShowingSettingsDialog();
 }
 
+bool Application::eventFilter(QObject *watched, QEvent *event)
+{
+    if (watched && watched->isWidgetType()
+        && (event->type() == QEvent::Show || event->type() == QEvent::Hide)) {
+        auto dialog = qobject_cast<QDialog*>(watched);
```

Comment:
This might not be a dialog and we might have windows which are not dialogs... we even got one which isn't a widget (the systray one) so that might miss a few cases.

For what you're trying to achieve, I'd look for `watcher->isWindowType()` but also `watcher->isWidgetType()`, and if that's a widget type I'd downcast to `QWidget` and check `isWindow()` there. That would catch anything which is top level.

Reply:
Doing this with `watched->isWindowType() && watched->isWidgetType()` didn't work (false) and using only one of them triggered the `Q_ASSERT(dialog)` in `ownCloudGui::updateDialogVisibility`.

What worked:
```
    if ((event->type() == QEvent::Show || event->type() == QEvent::Hide)
        && watched && watched->isWidgetType()) {
        auto widget = qobject_cast<QWidget*>(watched);

        if (widget->isWindow() && widget->isTopLevel()) {
            auto dialog = qobject_cast<QDialog*>(widget);
```

However this triggers the assert again when e.g. the Log window is closed. Since I actually don't like the agnostic filter approach in Application like this for such reasons: What do you think about installing the filters in the specific dialog constructors only for desired dialogs like Settings and sharing?

Reply:
Honestly I'd be leaning against moving it per dialog because it'd mean less separation of concerns and more duplicated code.

Reply:
> Doing this with `watched->isWindowType() && watched->isWidgetType()` didn't work (false) and using only one of them triggered the `Q_ASSERT(dialog)` in `ownCloudGui::updateDialogVisibility`.

Well yes, `watched->isWindowType() && watched->isWidgetType()` can never be true (it's either inheriting from QWidget or from QWindow but not both). In case that was unclear when I said "but also" I didn't have a logical and in mind. :-)

Due to the nature of the inheritance tree you basically need to manage two lists: the one which holds QDialogs and the one which holds QWindows.
 
> What worked:
> 
> ```
>     if ((event->type() == QEvent::Show || event->type() == QEvent::Hide)
>         && watched && watched->isWidgetType()) {
>         auto widget = qobject_cast<QWidget*>(watched);
> 
>         if (widget->isWindow() && widget->isTopLevel()) {
>             auto dialog = qobject_cast<QDialog*>(widget);
> ```

For the reasons above it'll never get in that second if AFAIK.